Implement `AnimationMixer` as a base class of `AnimationPlayer` and `AnimationTree`

TokageItLab opened the pull request:
Implement https://github.com/godotengine/godot-proposals/issues/5972.

Currently AnimationPlayer and AnimationTree have separate duplicate implementations of playback and blending, so implement the base class to unify them.

Internally, this is a major change, but most compatibility will be kept.

However, Changes to conflicting method and property names between AnimationPlayer and AnimationTree are unavoidable, since retaining only one of them will destroy compatibility with the other.

- `AnimationPlayer.playback_active`
- `AnimationTree.active`
  - -> `AnimationMixer.active` (but not change the setter/getter method since AnimationPlayer has already have the `AnimationPlayer.set_active()`  as setter of `AnimationPlayer.playback_active`)
- `AnimationPlayer.playback_process_mode`
- `AnimationTree.process_callback`
  - -> `AnimationMixer.callback_mode_process`
- `AnimationPlayer.method_call_mode`
  - -> `AnimationMixer.callback_mode_method` (to follow above change to make group in the inspector)

_Note: Old methods and property names will be kept for compatibility._
 
![image](https://github.com/godotengine/godot/assets/61938263/1ee7e51e-c6c0-46f6-b7c5-6250b98c4de7)
 
This change allows AnimationTree to have an AnimationLibrary. Also, when an AnimationTree is selected, a dummy AnimationPlayer is added to the editor internally.

![image](https://github.com/godotengine/godot/assets/61938263/e56ccc33-d196-4005-8962-36266075fe78)

If the AnimationTree specifies an AnimationPlayer, the dummy AnimationPlayer's resources will be referenced specified AnimationPlayer (make them read-only in AnimationTree).

![image](https://github.com/godotengine/godot/assets/61938263/4b97fa15-eaed-4d34-b661-7a58ed8d0673)

Todo:
  - [x] Implement AnimationMixer to unify blend methods
  - [x] Control `active` property between dummy player and selected node
  - [x] Implement blending algorithm which is not needed for RESET tracks for compatibility with old AnimationPlayer blending: Added `Deterministic` option, see [comment](https://github.com/godotengine/godot/pull/80813#issuecomment-1720424179)
  - [x] New/Re-implement "reset on save" for AnimationMixer (as mentioned in https://github.com/gdquest-demos/godot-4-3d-third-person-controller/pull/30#issue-1855201252)
  - [ ] ~Rework AnimationPlaybackTrack to fix https://github.com/godotengine/godot/issues/70338~ Will do it later as a separate PR

---

- Closes https://github.com/godotengine/godot/issues/82135
- Closes https://github.com/godotengine/godot/issues/80971
- Closes https://github.com/godotengine/godot/issues/73571
- Closes https://github.com/godotengine/godot/issues/70207
- Closes https://github.com/godotengine/godot/issues/79020
- Closes https://github.com/godotengine/godot/issues/70123
- Closes https://github.com/godotengine/godot/issues/75197

TokageItLab committed:
Implement AnimationManager the base class of AnimationPlayer/Tree

TokageItLab commented:
For compatibility, the old setters and getters have been added. It means that some internal properties will have two property names, setters and getters.

@reduz has rejected this and suggested implementing separate property names, setters and getters for AnimationPlayer and AnimationTree, but I disagree strongly with that.

It would result in missing properties during interconversion between AnimationPlayer and AnimationTree. This causes problems with the "Change type" option of the SceneTreeDock and the creation of the DummyPlayer. So we must have at least one shared property name.

For that, @reduz suggested putting an option in AnimationPlayerEditor, but honestly it's confusing and there is no need to ignore SceneTreeDock.

Having two property names, setters and getters is certainly redundant, but I argue that redundancy is far better than going the inconvenient way and having a method that only those who have read the manuals can understand.

YuriSizov commented:
Just to be clear, this is still a breaking compatibility change because the inheritance chain is modified. This may or may not be an issue for 3rd party language bindings.

TokageItLab commented:
Implemented "Predictable" option to allows blending without the RESET animation. It is `false` by default in `AnimationManager` and `AnimationPlayer`, and `true` by default in `AnimationTree`

![image](https://github.com/godotengine/godot/assets/61938263/a6c9a2f3-033c-4635-9a49-f4ad9b2054f3)

It normalize the total weight of the blend to be always 0 or 1, so it has the limitation that Add2 cannot be used (as the same in Godot 3). There is a PR like https://github.com/godotengine/godot/pull/76310, but it cannot coexist in Godot 4 because it relies on the fact that the blending in Godot 3 is not normalized and broken.

AThousandShips commented:
Add the compatibility breaking notes at the end of the file, and keep two lines between it and the previous entry

AThousandShips commented:
(At the end of the file, as per the instruction just above :) )

dsnopek commented:
Regarding GDExtension compatibility:

I think adding a new base class may be OK. Looking at `ClassDB::get_method_with_compatibility()`, it will walk up the inheritance chain and should get methods with the same name on the parent. GDExtension won't know what to do with an `AnimationManager` object if one is passed to it, but presumably GDExtension would only interact with `AnimationPlayer` and `AnimationTree` objects, so that should be fine.

I'm less clear on what's going on with the method renames and signature changes, because there's just so many changes in the diff. But anywhere that a method was "renamed", you'll need to make sure the old method is still there and registered, but calling the new method (so, it's really more like the old method was deprecated). I think the old deprecated methods can be registered on either the parent or child class (based on what I saw above in `ClassDB::get_method_with_compatibility()`).

And if any method signature changes, you'll need to add a compatibility method. These entries in `4.1-stable.expected` are suspicious:

```
Validate extension JSON: Error: Hash changed for 'classes/AnimationPlayer/methods/seek', from 7C72B6AA to 6BC1F2AB. This means that the function has changed and no compatibility function was provided.
Validate extension JSON: Error: Hash changed for 'classes/AnimationTree/methods/get_animation_player', from F2E72D3B to 1083D8C6. This means that the function has changed and no compatibility function was provided.
Validate extension JSON: Error: Hash changed for 'classes/AnimationTree/methods/get_tree_root', from 57256E6F to F4FF7E48. This means that the function has changed and no compatibility function was provided.
Validate extension JSON: Error: Hash changed for 'classes/AnimationTree/methods/set_tree_root', from 2A7D874F to 99E15E58. This means that the function has changed and no compatibility function was provided.
```

If the compatibility methods were setup correctly for the method signature changes, you wouldn't need to include those entries. I don't see the usual `_bind_compatibility_methods()` anywhere in the diff, so I think they are missing.

SaracenOne reviewed (approved):
Been out of the loop for a while, but want to get back into things. First read through and tests of this looks good. The abstractions seem sensible. 👍

TokageItLab commented on editor/plugins/animation_player_editor_plugin.cpp:
```diff
@@ -1979,14 +2052,89 @@ void AnimationPlayerEditorPlugin::_update_keying() {
+		if (!(E.usage & PROPERTY_USAGE_STORAGE)) {
+			continue;
+		}
+		if (!(E.name == "script" && E.name == "active" && E.name == "predictable" && E.name == "root_motion_track")) {
```
I think and chain is faster since it can detect false early.

AThousandShips replied:
This condition looks backwards, it will always be false, is it not supposed to be:
```suggestion
		if (!(E.name == "script" || E.name == "active" || E.name == "predictable" || E.name == "root_motion_track")) {
```
I.e.
```suggestion
		if (E.name != "script" && E.name != "active" && E.name != "predictable" && E.name != "root_motion_track") {
```

KoBeWi commented:
btw AnimationManager needs an icon.
CC @MewPurPur

MewPurPur commented:
If it's a non-instantiable node, I'd just use a grayed out AnimationPlayer icon. Is this the case?

TokageItLab commented:
![image](https://github.com/godotengine/godot/assets/61938263/c195bcf7-0438-44fb-886b-1b46680f6557)
Duplicated with AnimatedTexture... but I think okay.

MewPurPur commented:
So it is instantiable? You can add it to your SceneTree?

If so, I'm fine with this, can't really think of anything better.

TokageItLab commented:
@dsnopek Thanks for reviwe, I have addressed the four you pointed out.

- The const signature change was unintentional and has been revert.
- Added .compat.inc for `AnimationPlayer::seek()` and `AnimationTree::get/set_tree_root()`.
- For `AnimationManager::CallbackModeXXX`, even though .compat.inc was added, the header still needs to include the changed enum, so it was included directly in .cpp as deprecated

TokageItLab commented:
@MewPurPur Finally
![image](https://github.com/godotengine/godot/assets/61938263/6c143c42-b971-4c48-accf-d64c7e959c69)

TokageItLab commented:
Not so much related to this PR, but included a few minor fixes such as fadeout cache not resetting properly in AnimationPlayer blending, and current animation changes in the inspector not synchronizing with AnimationTrackEditor.

https://github.com/godotengine/godot/assets/61938263/e8c63943-75cb-471f-83ad-d0ad13d8be0b

dsnopek commented on misc/extension_api_validation/4.1-stable.expected:
```diff
@@ -169,10 +170,85 @@ Validate extension JSON: API was removed: classes/GraphNode/signals/raise_reques
+Validate extension JSON: Error: Field 'classes/AnimationPlayer/methods/seek/arguments': size changed value in new API, from 2 to 3.
+Validate extension JSON: Error: Field 'classes/AnimationTree/methods/get_tree_root/return_value': type changed value in new API, from "AnimationNode" to "AnimationRootNode".
+Validate extension JSON: Error: Field 'classes/AnimationTree/methods/set_tree_root/arguments/0': type changed value in new API, from "AnimationNode" to "AnimationRootNode".
+Validate extension JSON: Error: Hash changed for 'classes/AnimationPlayer/methods/seek', from 7C72B6AA to E68A17D4. This means that the function has changed and no compatibility function was provided.
+Validate extension JSON: Error: Hash changed for 'classes/AnimationTree/methods/get_tree_root', from 57256E6F to F4FF7E48. This means that the function has changed and no compatibility function was provided.
```
These two exceptions are still here. If they are no longer needed (because of the compat methods), then please remove them. If they are still needed, it probably means something isn't quite right with the compat methods.

TokageItLab replied:
Oops, it has lacked a DEFVAL and a class mismatch (it was returning a child class). I think it is no longer a problem now.

TokageItLab commented:
Renamed the `Predictable` option to `Deterministic`, which would be more suitable terminologically.

KoBeWi commented on scene/animation/animation_manager.cpp:
```diff
@@ -0,0 +1,2029 @@
+
+	ADD_SIGNAL(MethodInfo("animation_list_changed"));
+	ADD_SIGNAL(MethodInfo("animation_libraries_updated"));
+	ADD_SIGNAL(MethodInfo("animation_changed", PropertyInfo(Variant::STRING_NAME, "old_name"), PropertyInfo(Variant::STRING_NAME, "new_name")));
```
Why is this defined in AnimationManager if only AnimationPlayer ever uses it?

TokageItLab replied:
Indeed, I was confused with AnimationLibrary::animation_changed. Maybe it should be renamed something like AnimationPlayer::queue_advanced, but for now I've moved it to AnimationPlayer with keeping the current name.

KoBeWi commented:
Found 2 bugs with RESET track:
- when you save the scene, the animation is reset in the editor. It should not be happening, RESET should only affect what is saved

https://github.com/godotengine/godot/assets/2223172/ce44bdad-f3fb-4031-877a-c68dc1540f05

- when you create a keyframe and RESET animation does not exist, animation editor will switch to it when it's created

https://github.com/godotengine/godot/assets/2223172/fffdd6ee-ceb7-461a-9b1a-f906b2311c06

TokageItLab commented:
@KoBeWi Thank you for testing, I will fix later:

> when you save the scene, the animation is reset in the editor. It should not be happening, RESET should only affect what is saved

Maybe caused by the fact it restores the track_cache after a reset, but after end of playback there is no track_cache. For AnimationPlayer, I think it is possible to seek to the current position with a virtual function, but I think there is no way to do this for AnimationTree.

> when you create a keyframe and RESET animation does not exist, animation editor will switch to it when it's created

I think this is because the animation library has been updated and the TrackEditor has been initialized. And I guess this can be fixed by adding a function to make the Editor select the current name again.

TokageItLab commented:
> tbh I once wondered what's the point of this duplicate functionality.

I remember getting different results when blending, with the difference being that the value is only updated at the moment the actual key is crossed or always applied.

TokageItLab commented:
Maybe it should be all animations iterated and changed if the update mode is changed. I remember that issue has been raised some time somewhere.

TokageItLab commented:
@KoBeWi Alright, I believe the bug about RESET animations has been fixed.

As for blending different update modes, for now I have changed it to avoid spamming errors and suggest using Continuous with Nearest. I would like to follow up on the usability issue in a future PR.

TokageItLab commented:
Now for the blending different update modes, the Discrete animation to be treated as Continuous+Nearest internally when blending with a warning.

KoBeWi commented on doc/classes/AnimatableBody2D.xml:
```diff
@@ -4,14 +4,14 @@
 	<members>
 		<member name="sync_to_physics" type="bool" setter="set_sync_to_physics" getter="is_sync_to_physics_enabled" default="true">
-			If [code]true[/code], the body's movement will be synchronized to the physics frame. This is useful when animating movement via [AnimationPlayer], for example on moving platforms. Do [b]not[/b] use together with [method PhysicsBody2D.move_and_collide].
+			If [code]true[/code], the body's movement will be synchronized to the physics frame. This is useful when animating movement via [AnimationManager], for example on moving platforms. Do [b]not[/b] use together with [method PhysicsBody2D.move_and_collide].
```
I wonder if changing AnimationPlayer references to AnimationManager makes sense everywhere. I think things like moving platforms aren't usually made using AnimationTree.

KoBeWi commented on editor/animation_track_editor.cpp:
```diff
@@ -6303,6 +6313,18 @@ void AnimationTrackEditor::_show_imported_anim_warning() {
+void AnimationTrackEditor::_show_inactive_player_warning() {
+	EditorNode::get_singleton()->show_warning(
+			TTR("AnimationPlayer is inactive. The playback result is not applied to the scene."),
+			TTR("Warning: AnimationPlayer is inactive"));
```
```suggestion
			TTR("Warning"));
```
I think Warning is sufficient, as the same thing is said in the message.
Same above.

KoBeWi commented on editor/gui/scene_tree_editor.cpp:
```diff
@@ -115,7 +115,7 @@ void SceneTreeEditor::_cell_button_pressed(Object *p_item, int p_column, int p_i
 		undo_redo->commit_action();
 	} else if (p_id == BUTTON_PIN) {
-		if (n->is_class("AnimationPlayer")) {
+		if (n->is_class("AnimationPlayer") || n->is_class("AnimationTree") || n->is_class("AnimationManager")) {
```
```suggestion
		if (n->is_class("AnimationManager")) {
```
`is_class()` already checks for base class.

KoBeWi commented on editor/gui/scene_tree_editor.cpp:
```diff
@@ -465,8 +465,8 @@ void SceneTreeEditor::_add_nodes(Node *p_node, TreeItem *p_parent) {
 			_update_visibility_color(p_node, item);
-		} else if (p_node->is_class("AnimationPlayer")) {
-			bool is_pinned = AnimationPlayerEditor::get_singleton()->get_player() == p_node && AnimationPlayerEditor::get_singleton()->is_pinned();
+		} else if (p_node->is_class("AnimationPlayer") || p_node->is_class("AnimationTree") || p_node->is_class("AnimationManager")) {
```
```suggestion
		} else if (p_node->is_class("AnimationManager")) {
```
Same.

KoBeWi commented on editor/plugins/animation_state_machine_editor.cpp:
```diff
@@ -92,7 +92,7 @@ String AnimationNodeStateMachineEditor::_get_root_playback_path(String &r_node_d
 		while (!is_playable_anodesm_found) {
 			base_path = String("/").join(edited_path);
-			Ref<AnimationNodeStateMachine> anodesm = !edited_path.size() ? tree->get_tree_root() : tree->get_tree_root()->find_node_by_path(base_path);
+			Ref<AnimationNodeStateMachine> anodesm = !edited_path.size() ? Ref<AnimationNode>(tree->get_root_animation_node().ptr()) : tree->get_root_animation_node()->find_node_by_path(base_path);
```
```suggestion
			Ref<AnimationNodeStateMachine> anodesm = !edited_path.size() ? tree->get_root_animation_node() : tree->get_root_animation_node()->find_node_by_path(base_path);
```
Why the weird conversion? Is it to silence some warning?

TokageItLab replied:
Yes, get_root_animation_node() now only returns RootNode, so it will give a warning that the classes of the ternary operators do not match.

KoBeWi commented on scene/animation/animation_manager.h:
```diff
@@ -0,0 +1,389 @@
+
+		real_t total_weight = 0.0;
+
+		TrackCache() {
+		}
```
```suggestion
		TrackCache() { }
```
Though empty constructor and destructor is likely unnecessary.

KoBeWi commented on scene/animation/animation_player.cpp:
```diff
@@ -199,816 +131,35 @@ void AnimationPlayer::_validate_property(PropertyInfo &p_property) const {
-
 	for (const KeyValue<StringName, AnimationData> &E : animation_set) {
-		if (E.value.next != StringName()) {
+		if (animation_next_set.has(E.key) && animation_next_set[E.key] != StringName()) {
```
```suggestion
		HashMap<StringName, StringName>::ConstIterator F = animation_next_set.find(E.key);
		if (F && F->value != StringName()) {
```
Avoids double lookup.

KoBeWi commented on editor/plugins/animation_player_editor_plugin.cpp:
```diff
@@ -1007,7 +1043,31 @@ void AnimationPlayerEditor::_update_name_dialog_library_dropdown() {
+		}
+	}
+
+	// Make some options disable.
```
```suggestion
	// Make some options disabled.
```

KoBeWi commented on editor/plugins/animation_player_editor_plugin.cpp:
```diff
@@ -1007,7 +1043,31 @@ void AnimationPlayerEditor::_update_name_dialog_library_dropdown() {
 
-void AnimationPlayerEditor::edit(AnimationPlayer *p_player) {
+void AnimationPlayerEditor::_ensure_dummy_player() {
+	bool is_dummy_player_existed = is_dummy && player && original_node;
```
```suggestion
	bool dummy_exists = is_dummy && player && original_node;
```
`is_dummy_player_existed` is grammatically incorrect.

KoBeWi reviewed (approved):
I tested some basic functionality. The bugs I found before are fixed, I didn't find anything else. I did not touch AnimationTree though, I think it might need some testing to find obvious bugs (unless you already tested it?).

I skimmed through the code and caught some random things, but there is so many changes and most of it looks like copy-paste. I think overall it's fine.

lyuma commented:
I tested it quite a bit and it's working correctly in all of my test scenes.

I found turning the "Deterministic" checkbox off was helpful for emulating the behavior some Unity animation controllers, where state is expected to be preserved when transitioning states (similar to the behavior of Unity's "Write Defaults" turned off.

I did encounter one crash when doing a combination of opening and closing scenes and selecting the AnimationTree and AnimationPlayer nodes in some order. Unfortunately I was unable to reproduce it by following the same steps leading up to the crash.
It's also very likely that this crash existed before AnimationManager: I've had AnimationPlayer crash before, for example. So it might not be a regression.
```
================================================================
CrashHandlerException: Program crashed
Engine version: Godot Engine v4.2.dev.custom_build (59139df16e7a10c3b9176f697d23b557af46601e)
Dumping the backtrace. Please include this when reporting the bug to the project developer.
[0] Object::set (S:\repo\godot-pr\core\object\object.cpp:228)
[1] AnimationPlayerEditorPlugin::_update_dummy_player (S:\repo\godot-pr\editor\plugins\animation_player_editor_plugin.cpp:2166)
[2] CallableCustomMethodPointer<AnimationPlayerEditorPlugin,AnimationManager *>::call (S:\repo\godot-pr\core\object\callable_method_pointer.h:104)
[3] Callable::callp (S:\repo\godot-pr\core\variant\callable.cpp:51)
[4] CallableCustomBind::call (S:\repo\godot-pr\core\variant\callable_bind.cpp:145)
[5] Callable::callp (S:\repo\godot-pr\core\variant\callable.cpp:51)
[6] CallQueue::_call_function (S:\repo\godot-pr\core\object\message_queue.cpp:220)
[7] CallQueue::flush (S:\repo\godot-pr\core\object\message_queue.cpp:340)
[8] SceneTree::physics_process (S:\repo\godot-pr\scene\main\scene_tree.cpp:473)
[9] Main::iteration (S:\repo\godot-pr\main\main.cpp:3488)
[10] OS_Windows::run (S:\repo\godot-pr\platform\windows\os_windows.cpp:1474)
[11] widechar_main (S:\repo\godot-pr\platform\windows\godot_windows.cpp:182)
[12] _main (S:\repo\godot-pr\platform\windows\godot_windows.cpp:206)
[13] main (S:\repo\godot-pr\platform\windows\godot_windows.cpp:218)
[14] __scrt_common_main_seh (D:\a\_work\1\s\src\vctools\crt\vcstartup\src\startup\exe_common.inl:288)
[15] <couldn't map PC to fn name>
-- END OF BACKTRACE --
================================================================
```

One slight usability regression is that the "Active" checkbox for AnimationTree is now part of the AnimationManager category, which requires scrolling past all of the parameters to toggle.

Overall this change looks extremely good. I didn't read every line of code but the overall organization seems good.

KoBeWi commented:
> It's also very likely that this crash existed before AnimationManager: I've had AnimationPlayer crash before, for example. So it might not be a regression.

The `_update_dummy_player()` was added in this PR. Looking at the code, seems like the player was freed without resetting the pointer.

lyuma commented:
I'm having a problem with using a GDExtension designed for Godot 4.1.1 on this build:
```
ERROR: Rust function panicked in file C:\Users\crann\.cargo\git\checkouts\gdext-46fb6d0c2ad8264d\4e249d3\godot-ffi\src\toolbox.rs at line 201. Context: failed to initialize GDExtension level `Scene`
   at: <function unset> (C:\Users\crann\.cargo\git\checkouts\gdext-46fb6d0c2ad8264d\4e249d3\godot-core\src\lib.rs:151)
ERROR: Panic msg:
  Failed to load class method AnimationPlayer::set_root (hash 1348162250).
  Make sure gdext and Godot are compatible: https://godot-rust.github.io/book/gdext/advanced/compatibility.html
```

TokageItLab commented:
@lyuma Thank you for testing, I added a compat method to AnimationPlayer. Also I added an error check in _update_dummy_player where it looked dangerous, but I am not sure if that is really the cause.
If it doesn't happen often, it is probably not critical so I would like to address it when I have a procedure to reproduce the crash and the cause later.

dsnopek commented:
> Thank you for testing, I added a compat method to AnimationPlayer.

With how the system works currently, you can't add a compat method for a method that doesn't exist. You have to bind a normal (but deprecated) method with the old name that calls the new method. Compat methods are only for method signature changes.

(Of course, if we wanted to, we could update the system so that you _can_ have compat methods for methods that doesn't exist, but that's not how it works currently.)

TokageItLab commented:
Now it uses `bind_method` instead of `bind_compatibility_method`.

It is needed to rename because during the base class integration, we now have a method called `set_root_motion_track` in addition to `set_root`, which makes the meaning unclear.

I think the other methods are correctly implemented in BaseClass, so there should be no other problems.

lyuma commented:
```
ERROR: Method 'AnimationPlayer.set_process_callback' has changed and no compatibility fallback has been provided. Please open an issue.
   at: gdextension_classdb_get_method_bind (core\extension\gdextension_interface.cpp:1310)
ERROR: Rust function panicked in file C:\Users\crann\.cargo\git\checkouts\gdext-46fb6d0c2ad8264d\4e249d3\godot-ffi\src\toolbox.rs at line 201. Context: failed to initialize GDExtension level `Scene`
   at: <function unset> (C:\Users\crann\.cargo\git\checkouts\gdext-46fb6d0c2ad8264d\4e249d3\godot-core\src\lib.rs:151)
ERROR: Panic msg:
  Failed to load class method AnimationPlayer::set_process_callback (hash 1663839457).
  Make sure gdext and Godot are compatible: https://godot-rust.github.io/book/gdext/advanced/compatibility.html
   at: <function unset> (C:\Users\crann\.cargo\git\checkouts\gdext-46fb6d0c2ad8264d\4e249d3\godot-core\src\lib.rs:99)
```

In case you would like to test...
The project I'm using is https://github.com/virtual-puppet-project/vpuppr
It uses a godot-rust GDExtension module compiled by https://github.com/virtual-puppet-project/libvpuppr

TokageItLab commented:
@lyuma Does the problem occur when compiling gd-rust? (I don't have the rust environment so I couldn't test it)

At any rate, for `set/get_process_callback()` and `set/get_method_call_mode()`, there were ENUM and signature changes, so now I added those to the base class compat.inc.

lyuma commented:
The hash still changed, so it does not work yet.

Godot 4.2 AnimationManager.set_process_callback:
```json
{
        "name": "set_process_callback",
        "is_const": false,
        "is_vararg": false,
        "is_static": false,
        "is_virtual": false,
        "hash": 299281246,
        "arguments": [
                {
                        "name": "mode",
                        "type": "enum::AnimationManager.AnimationProcessCallback"
                }
        ]
},
```

Godot 4.1.1 AnimationPlayer.set_process_callback:
```json
{
        "name": "set_process_callback",
        "is_const": false,
        "is_vararg": false,
        "is_static": false,
        "is_virtual": false,
        "hash": 1663839457,
        "arguments": [
                {
                        "name": "mode",
                        "type": "enum::AnimationPlayer.AnimationProcessCallback"
                }
        ]
},
```

Godot 4.1.1 AnimationTree.set_process_callback:
```json
{
        "name": "set_process_callback",
        "is_const": false,
        "is_vararg": false,
        "is_static": false,
        "is_virtual": false,
        "hash": 1723352826,
        "arguments": [
                {
                        "name": "mode",
                        "type": "enum::AnimationTree.AnimationProcessCallback"
                }
        ]
},
```

The issue is likely that the hash of the enum class name changed.

The method `MethodBind::get_hash()` in method_bind.cpp performs the following hash on the arguments and return values:
```cpp
	for (int i = (has_return() ? -1 : 0); i < get_argument_count(); i++) {
		PropertyInfo pi = i == -1 ? get_return_info() : get_argument_info(i);
		hash = hash_murmur3_one_32(get_argument_type(i), hash);
		if (pi.class_name != StringName()) {
			hash = hash_murmur3_one_32(pi.class_name.operator String().hash(), hash);
		}
	}
```
This means any attempt to rename, move or unify an enum will break compatibility in this way.

I think this is a problem that needs to be discussed with the GDExtension team. My personal opinion is sub-types like enum vs int or float vs double should not change the hash, since the underlying data type passed into the function would be the same at an ABI level.

TokageItLab commented:
So in short, instead of adding an Enum to BaseClass, do we need to add an Enum for compatibility to the individual child classes? Hmm...

TokageItLab commented:
In the meantime, I added a compatibility enum to each of the child classes. Although it looks messy...

MewPurPur commented on editor/icons/AnimationManager.svg:
```diff
@@ -0,0 +1 @@
+<svg height="16" viewBox="0 0 16 16" width="16" xmlns="http://www.w3.org/2000/svg"><path d="M1 1v14h1v-2h2v2h8v-2h2v2h1V1h-1v2h-2V1H4v2H2V1zm1 4h2v2H2zm10 0h2v2h-2zM2 9h2v2H2zm10 0h2v2h-2z" fill="#919191"/></svg>
```
```suggestion
<svg height="16" viewBox="0 0 16 16" width="16" xmlns="http://www.w3.org/2000/svg"><path d="M1 1v14h1v-2h2v2h8v-2h2v2h1V1h-1v2h-2V1H4v2H2V1zm1 4h2v2H2zm10 0h2v2h-2zM2 9h2v2H2zm10 0h2v2h-2z" fill="#808080"/></svg>
```

TokageItLab replied:
NodeDisabled uses #919191. view-source:https://raw.githubusercontent.com/godotengine/godot/master/editor/icons/NodeDisabled.svg

Bromeon commented:
_Regarding the compatibility policy (can be moved to another issue if needed):_

Unfortunately, changing the policy to "enums are compatible" now would mean re-hashing _all_ methods that work with enums. This would need 2 orders of magnitude more compat methods 😕 

Manually adding compat methods here is one action. But if we expect more such renamings, it makes sense to look for other strategies.

One option would be to maintain a list of renamed types. So if you have the `A -> B` renaming in that list, and the method registration system encounters a method with type `B`, it will auto-register the same method a 2nd time, but computes the hash as if `A` were used in place of `B`. The effect is the same as manually registering all compat methods, but instead of dozens of registrations, you only have it once in code.

lyuma reviewed (commented):
> In the meantime, I added a compatibility enum to each of the child classes. Although it looks messy...

tested the PR and it gets past AnimationPlayer now, so this does look like it solved the issue here even if messy

lyuma commented on scene/animation/animation_player.compat.inc:
```diff
@@ -0,0 +1,76 @@
+	ClassDB::bind_method(D_METHOD("set_root", "path"), &AnimationPlayer::_set_root_bind_compat_80813);
+	ClassDB::bind_method(D_METHOD("get_root"), &AnimationPlayer::_get_root_bind_compat_80813);
+	ClassDB::bind_compatibility_method(D_METHOD("seek", "seconds", "update"), &AnimationPlayer::_seek_bind_compat_80813, DEFVAL(false));
+	BIND_ENUM_CONSTANT(ANIMATION_PROCESS_PHYSICS);
+	BIND_ENUM_CONSTANT(ANIMATION_PROCESS_IDLE);
+	BIND_ENUM_CONSTANT(ANIMATION_PROCESS_MANUAL);
+	BIND_ENUM_CONSTANT(ANIMATION_METHOD_CALL_DEFERRED);
+	BIND_ENUM_CONSTANT(ANIMATION_METHOD_CALL_IMMEDIATE);
```
I tested commenting out the BIND_ENUM_CONSTANT and it seems to avoid adding the unneeded constant documentation to AnimationPlayer and AnimationTree at least in godot-rust

lyuma replied:
Though... technically speaking, a GDScript user might have referenced `AnimationPlayer.ANIMATION_PROCESS_IDLE` in Godot 4.1, and so having these enum constants would help keep compatibility with GDScript code.

So maybe best not to remove the exposed enum constants for script compat.

TokageItLab replied:
Well, I will just leave it there for now. Thank you for the testing.

TokageItLab commented:
I have tested more and fixed some parts that were causing freezes and crashes. I think it is almost ready to be merged.

@reduz Please confirm this as a final approval. I believe this base class design is exactly as we agreed previously.

AThousandShips commented on editor/plugins/animation_player_editor_plugin.cpp:
```diff
@@ -1979,14 +2102,88 @@ void AnimationPlayerEditorPlugin::_update_keying() {
+	// Add dummy player to scene.
+	if (!dummy_player) {
+		Node *parent = p_manager->get_parent();
+		ERR_FAIL_COND(!parent);
```
```suggestion
		ERR_FAIL_NULL(parent);
```

TokageItLab reviewed (commented):
@reduz To make the review easier, I picked up a few main points of change.

TokageItLab commented:
AnimationManager has been renamed to AnimationMixer by discussion on RocketChat: we decided to name this base class AnimationMixer.

![image](https://github.com/godotengine/godot/assets/61938263/f4ed6eb4-3a6d-4963-9925-37b978fa5e17)

YuriSizov commented on scene/animation/animation_player.h:
```diff
@@ -394,46 +148,42 @@ class AnimationPlayer : public Node {
-	bool can_apply_reset() const;
-#endif
-
+#ifndef DISABLE_DEPRECATED
+	enum AnimationProcessCallback {
+		ANIMATION_PROCESS_PHYSICS,
+		ANIMATION_PROCESS_IDLE,
+		ANIMATION_PROCESS_MANUAL,
+	};
+	enum AnimationMethodCallMode {
+		ANIMATION_METHOD_CALL_DEFERRED,
+		ANIMATION_METHOD_CALL_IMMEDIATE,
+	};
+	void _set_process_callback_bind_compat_80813(AnimationProcessCallback p_mode);
+	AnimationProcessCallback _get_process_callback_bind_compat_80813() const;
+	void _set_method_call_mode_bind_compat_80813(AnimationMethodCallMode p_mode);
+	AnimationMethodCallMode _get_method_call_mode_bind_compat_80813() const;
+	void _set_root_bind_compat_80813(const NodePath &p_root);
+	NodePath _get_root_bind_compat_80813() const;
+	void _seek_bind_compat_80813(double p_time, bool p_update = false);
+
+	static void _bind_compatibility_methods();
+#endif // DISABLE_DEPRECATED
```
These shouldn't be public (same for the animation tree).

TokageItLab replied:
ENUM seems to need to be public, but otherwise moved.

YuriSizov commented on scene/animation/animation_player.h:
```diff
@@ -394,46 +148,42 @@ class AnimationPlayer : public Node {
 	~AnimationPlayer();
 };
-
+#ifndef DISABLE_DEPRECATED
 VARIANT_ENUM_CAST(AnimationPlayer::AnimationProcessCallback);
 VARIANT_ENUM_CAST(AnimationPlayer::AnimationMethodCallMode);
+#endif // DISABLE_DEPRECATED
```
And would be nice to preserve the line break before this block (same for the animation tree).

YuriSizov commented:
Looks great, just a small style change, if you could.

akien-mga commented:
Thanks!

RandomShaper commented:
After rebasing #80939 on top of this, onion skinning can no longer work because backing up and restoring animated values now requires a reset animation to be present. That didn't use to be a limitation. And, even if there happens to be a reset animation, I'm not sure if onion skinning would behave the same as it did. @TokageItLab, can you give me a hint on how could I capture the current animated values with the new API, to, say, create a bespoke reset animation for onion skinning purposes? Alternatively, I could avoid enabling onion skinning unless there's a reset animation present, but I'm not very fond of that.

lyuma commented:
@RandomShaper From the testing I did, I was generally able to use the Deterministic checkbox to control whether it follows the old rules for AnimationPlayer (keeping all existing animations) or the original rules for AnimationTree (restoring the animated values).

Would it be possible to test both Determinstic on and off and see if one has behavior closer to what you're looking for?

From what I skimmed, onion skinning uses AnimationPlayer, so I would expect Deterministic = off should work similar to before.

TokageItLab commented:
Onion skin was disabled as an experimental/broken feature during AnimationManager testing, so we could not confirm its correct behavior.

I remember that saw the reset/restore function used there, but I have not touched it so much.

I have rewritten the reset/restore function itself and changed the reset_on_save behavior, so the onion skin may need to be changed to follow/match that change. Would copy-pasting a line similar to the code I rewrote in the reset_on_save change be sufficient for it to work?

RandomShaper commented:
The problem is that my original backup/restore implementation was able to work without a reset animation and, as far as I can see, that is gone. I surely understand that happened because there was no (enabled) use case for it.

Now, for the time being, I'll have to make onion skinning require a reset animation. Hopefully someone (maybe myself) will be able to lift that requirement by bringing reset-less backup back.

TokageItLab commented:
I remember about reset_on_save, if there is no reset animation, I think it is made to use the initial (mostly zero) value to avoid updating the scene file. If you disable reset_on_save, does that mean that onion skin will work? And for onion skin, does it mean that reset_on_save is not available?

RandomShaper commented:
Unless it changed after I implemented it, reset-on-save would only work in `AnimationPlayer`s having a reset animation. Without such an animation, the feature is no-op.

The problem with onion skinning is that it works like this:
- Save all current values for every track and the current position (time).
- For every step, seek to the relevant time, render and take screenshot.
- Seek back to the original position and restore the backup.

It didn't require a reset animation, nor it it fitting for this purpose the way I envisioned it, because the current values in the animated properties may not even be the ones in the animation. For instance, say you have a `Node2D` animated such as its X position moves from 0 to 100 in some track. The current animation at the current time sets position to 50. The user drags the node manually to 15 and then enables onion skinning. After all the process involved (done on every frame), the user wants to still see the position is 15. The fact that seeking the `AnimationPlayer` is being used under the hood to capture the onion layers is kept hidden and unnoticed to the user.

With the reset animation there are a couple of downsides:
1. Well, you need to have a proper reset animation involving all the values that you want to behave correctly in the onion layers.
2. The onion skinning internal machinery is no longer hidden, since, the second you enable it, properties the user may have modified manually will be reset to the values mandated by the animation.

Neither is the end of the world, and I think that is good enough to bring onion skinning back to life. However, as I said, I'd like it to work as transparently (no pun intended) as it used to do in a future iteration.
